compiler/rustc_hir_typeck/src/expr.rs

```diff
@@ -1458,6 +1456,27 @@ impl<'a, 'tcx> FnCtxt<'a, 'tcx> {
         }
     }
 
+    fn check_expr_const_block(
+        &self,
+        block: &'tcx hir::Expr<'tcx>,
+        hir_id: HirId,
+        span: Span,
+        expected: Expectation<'tcx>,
+    ) -> Ty<'tcx> {
+        let feed = self.tcx().create_def(self.body_id, kw::Empty, DefKind::InlineConst);
+        feed.def_span(span);
+        feed.local_def_id_to_hir_id(hir_id);
```

Comment:
I feel like const blocks should be first-class nested bodies (liek closures) all the way from the AST-level (i.e. having a def id, being treated as a hir owner, etc) instead of having to synthesize a def-id here.

Reply:
that's what they used to be (before https://github.com/rust-lang/rust/pull/124650), but it meant we couldn't feed the `type_of` query. Turns out I couldn't do that anyway, so now I think I should probably just revert https://github.com/rust-lang/rust/pull/124650 and try again from scratch

Reply:
though we'll def need to lazily create the `DefId` of anon consts of method calls' const generic params, so that we can feed type_of. That's why I'm trying to figure out how to do this better.

Reply:
right now we are breaking query system invariants and are feeding anon const `type_of` from `typeck`, even though `typeck` did not create the `DefId`.

Reply:
Also, why do we need to feed the `type_of` query at all?

The `type_of` for a const block should operate like a `type_of` for a closure -- for closures, it calls `typeck_results` on the parent and extracts it from the `expr_ty`:
https://github.com/rust-lang/rust/blob/2a2c29aafa50bf6fe53d66b32070eba59f860ac3/compiler/rustc_hir_analysis/src/collect/type_of.rs#L483-L485

Reply:
> though we'll def need to lazily create the DefId of anon consts of method calls' const generic params, so that we can feed type_of. That's why I'm trying to figure out how to do this better.

I don't understand why this needs to be entangled with inline const exprs. Perhaps we should just distinguish these two rather than treating them the same.